### PR TITLE
Avoid allocations, fix CI

### PR DESCRIPTION
**Pull request by tschottdorf**
Embed StandardContext (not *StandardContext) into
RawSpan and initialize the attribute and tags map
only on the first write (it is valid to read from
a nil map).

Move the trace attributes from StandardContext onto
spanImpl; they were not exposed to the recorder, and
the lock on StandardContext required that object to
always be heap allocated. If they need to be exposed
to the user, they can be moved onto RawSpan (or even
back onto StandardContext, but without the lock).
Locking is instead synchronized through the Span's
log.

Work towards sane allocation behavior and add some
benchmarks to that effect: Span creation and destruction
with calls to LogEvent, SetTag oder SetTraceAttribute
now don't allocate.

Changed various other pointers to be passed by-value.

Fix various linter errors and the invocation in
the Makefile (which did not exit nonzero on issues).

Use a sync.Pool instead of new allocations …
This reduces the allocs for a common Span to zero.

```
name               old time/op    new time/op    delta
Span_Empty-8          411ns ± 0%     337ns ± 0%   ~     (p=1.000 n=1+1)
Span_100Events-8     18.2µs ± 0%    18.6µs ± 0%   ~     (p=1.000 n=1+1)
Span_1000Events-8    18.2µs ± 0%    19.3µs ± 0%   ~     (p=1.000 n=1+1)
Span_100Tags-8       25.1µs ± 0%    26.2µs ± 0%   ~     (p=1.000 n=1+1)
Span_1000Tags-8      24.7µs ± 0%    24.9µs ± 0%   ~     (p=1.000 n=1+1)

name               old alloc/op   new alloc/op   delta
Span_Empty-8           144B ± 0%       0B ±NaN%   ~     (p=1.000 n=1+1)
Span_100Events-8     16.2kB ± 0%    16.1kB ± 0%   ~     (p=1.000 n=1+1)
Span_1000Events-8    16.2kB ± 0%    16.1kB ± 0%   ~     (p=1.000 n=1+1)
Span_100Tags-8       10.7kB ± 0%    10.5kB ± 0%   ~     (p=1.000 n=1+1)
Span_1000Tags-8      10.7kB ± 0%    10.5kB ± 0%   ~     (p=1.000 n=1+1)

name               old allocs/op  new allocs/op  delta
Span_Empty-8           1.00 ± 0%     0.00 ±NaN%   ~     (p=1.000 n=1+1)
Span_100Events-8       9.00 ± 0%      8.00 ± 0%   ~     (p=1.000 n=1+1)
Span_1000Events-8      9.00 ± 0%      8.00 ± 0%   ~     (p=1.000 n=1+1)
Span_100Tags-8         12.0 ± 0%      11.0 ± 0%   ~     (p=1.000 n=1+1)
Span_1000Tags-8        12.0 ± 0%      11.0 ± 0%   ~     (p=1.000 n=1+1)
```

FYI, I'm piling more commits on top of this, but would be good to get this
in/discuss it before it gets more refined.

### CODE REVIEW

File: propagation.go [hunk 1]
```diff
@@ -7,19 +7,19 @@ import "errors"
 ///////////////////////////////////////////////////////////////////////////////
 
 var (
-	// TraceNotFound occurs when the `carrier` passed to Extractor.JoinTrace()
+	// ErrTraceNotFound occurs when the `carrier` passed to Extractor.JoinTrace()
 	// is valid and uncorrupted but has insufficient information to join or
 	// resume a trace.
-	TraceNotFound = errors.New("opentracing: Trace not found in Extraction carrier")
+	ErrTraceNotFound = errors.New("opentracing: Trace not found in Extraction carrier")
```
Comment (bensigelman) [inline propagation.go]: thanks for the renames

File: standardtracer/span.go [hunk 2]
```diff
@@ -11,22 +11,33 @@ import (
 // Implements the `Span` interface. Created via tracerImpl (see
 // `standardtracer.New()`).
 type spanImpl struct {
-	lock     sync.Mutex
-	tracer   *tracerImpl
-	recorder SpanRecorder
-	raw      RawSpan
+	tracer     *tracerImpl
+	sync.Mutex // protects the fields below
```
Comment (bensigelman) [inline standardtracer/span.go]: is there a reason you went for Mutex rather than RWMutex as before?

File: standardtracer/span.go [hunk 3]
```diff
@@ -11,22 +11,33 @@ import (
...
+	raw        RawSpan
+	// TODO(tschottdorf): should this be available to the Recorder
```
Comment (bensigelman) [inline standardtracer/span.go]: the recorder should have access to the trace attrs

File: span.go [hunk 4]
```diff
@@ -146,7 +146,7 @@ type FinishOptions struct {
 	//
 	// If specified, the caller hands off ownership of BulkLogData at
 	// FinishWithOptions() invocation time.
-	BulkLogData []*LogData
+	BulkLogData []LogData
```
Comment (yurishkuro) [inline span.go]: @bensigelman ^^^

File: standardtracer/span.go (hunk 2 above)
Comment (tschottdorf) [inline standardtracer/span.go]: `spanImpl` had a `Mutex` even before, and generally I would only use an `RWMutex` when it's indicated by a specific workload (since it has generally higher overhead). I would naively expect a trace to be "mostly" written to, so that a `Mutex` seems the better fit.

File: standardtracer/span.go (hunk 3 above)
Comment (tschottdorf) [inline standardtracer/span.go]: Ok. Don't think that was the case before, but I can fix it. I'll put it on RawSpan if that's ok.

File: propagation.go (hunk 1 above)
Comment (tschottdorf) [inline propagation.go]: thank `golint`, not me :-)

File: standardtracer/span.go [hunk 5]
```diff
@@ -45,14 +56,14 @@ func (s *spanImpl) LogEventWithPayload(event string, payload interface{}) {
...
-	defer s.lock.Unlock()
+	s.Lock()
+	defer s.Unlock()
 
 	if ld.Timestamp.IsZero() {
 		ld.Timestamp = time.Now()
 	}
 
-	s.raw.Logs = append(s.raw.Logs, &ld)
+	s.raw.Logs = append(s.raw.Logs, ld)
```
Comment (bensigelman) [inline standardtracer/span.go]: for large LogData instances I'm surprised that this is a perf improvement! but that's what benchmarks are for, I guess.

File: standardtracer/bench_test.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+	}
+}
+
+type countingRecorder int32
+
+func (c *countingRecorder) RecordSpan(r RawSpan) {
+	atomic.AddInt32((*int32)(c), 1)
+}
+
+func benchmarkWithOps(b *testing.B, numEvent, numTag, numAttr int) {
```
Comment (bensigelman) [inline standardtracer/bench_test.go]: thanks for adding this

File: standardtracer/span.go (hunk 2 above)
Comment (bensigelman) [inline standardtracer/span.go]: I'm fine with the decision, I was just curious if it came from a specific benchmark or more of an ad hoc hypothesis (sounds like the latter, and that's fine with me at this point). The RWMutex was there for the Trace Attributes specifically since I thought there might be more reads than writes there, but absent a contention profile I'm fine keeping it simple.

File: standardtracer/span.go (hunk 5 above)
Comment (tschottdorf) [inline standardtracer/span.go]: `LogData` is always small: a timestamp and two pointers.

File: standardtracer/bench_test.go (hunk 6 above)
Comment (tschottdorf) [inline standardtracer/bench_test.go]: I'm adding corresponding ones for the serializers right now.

File: standardtracer/span.go (hunk 5 above)
Comment (yurishkuro) [inline standardtracer/span.go]: I suppose because LogData is a timestamp and 2 pointers, faster to copy it than to allocate on heap. I.e. the speed improvement is probably coming from the points where LogData is created.

File: span.go (hunk 4 above)
Comment (bensigelman) [inline span.go]: Yes, I saw this... it's the only "interesting" (non-lint) change to the public API. I don't know whether it's reasonable to assume that the caller will always know how many `LogData`s they want to add at once. Nor do I know whether the above actually matters... It looks like @tschottdorf's benchmarks suggest the allocations are worse than the copying in the `RawSpan` case at least.

**Comment by bensigelman** [comment]
@tschottdorf the standardtracer stuff is not optimized (or at least it was not optimized) and I appreciate you taking the time to profile and fix it up. https://github.com/opentracing/opentracing-go/pull/53#discussion-diff-52850024 is the only concern I have... I wish we had a more realistic real-world benchmark. I'm okay proceeding with the non-pointer `[]LogData`, though, unless @yurishkuro or others (@slimsag?) with appreciable Go experience have words of warning to share on this front.

Basically LGTM, though, and thanks again!

File: standardtracer/span.go (hunk 5 above)
Comment (bensigelman) [inline standardtracer/span.go]: @tschottdorf @yurishkuro per https://github.com/opentracing/opentracing.github.io/issues/30, the LogData is going to include more things in the future (standard debug-logging fileds, namely).

File: standardtracer/span.go (hunk 3 above)
Comment (bensigelman) [inline standardtracer/span.go]: fine by me

File: standardtracer/bench_test.go (hunk 6 above)
Comment (bensigelman) [inline standardtracer/bench_test.go]: Thanks in advance for those, then.

I didn't want to do profiling and optimization until the API had stabilized... I guess we're there (or quite nearly there), so hopefully this isn't a waste of your time!

File: standardtracer/span.go (hunk 2 above)
Comment (tschottdorf) [inline standardtracer/span.go]: Yeah, no specific benchmark. I had two mutexes here and needed to decide for one.

File: standardtracer/span.go (hunk 5 above)
Comment (tschottdorf) [inline standardtracer/span.go]: like what? The only thing that would ruin it are many large fields. If we're adding more pointers (strings, ...) then the actual struct should remain fairly small.

File: span.go (hunk 4 above)
Comment (tschottdorf) [inline span.go]: > I don't know whether it's reasonable to assume that the caller will always know how many LogDatas they want to add at once.

Might be missing something, but how does that change change the game here?

**Comment by tschottdorf** [comment]
> I'm okay proceeding with the non-pointer []LogData, though, unless @yurishkuro or others (@slimsag?) with appreciable Go experience have words of warning to share on this front.

I'd be surprised if the pointer were better (we could always change it back). It's definitely worse in that unless you're tricky, each `LogEntry` is a single allocation in the previous version.

File: standardtracer/span.go (hunk 3 above)
Comment (tschottdorf) [inline standardtracer/span.go]: done.

File: standardtracer/bench_test.go (hunk 6 above)
Comment (tschottdorf) [inline standardtracer/bench_test.go]: I realize we're not quite there yet, but hopefully close to the final version. Benchmarking can expose flaws in the API as well, and I figured you'd be busy enough with all the fallout from stabilization, so I went to give Go a head start.

**Comment by yurishkuro** [comment]
lgtm

I would suggest merging incrementally than piling more commits onto this pr.

**Comment by tschottdorf** [comment]
Yep, I'm already holding off.
